### PR TITLE
perf(engine): parallel storage roots

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Enables parallel storage roots in consensus engine. Parallel calculation will be _skipped_ if persistence is currently in progress or if consistency check _fails_ during calculation, it will fall back to regular one

### CODE REVIEW

File: crates/engine/tree/src/tree/mod.rs [hunk 1]
```diff
@@ -1869,8 +1875,60 @@ where
         let hashed_state = HashedPostState::from_bundle_state(&output.state.state);
 
         let root_time = Instant::now();
-        let (state_root, trie_output) =
-            state_provider.state_root_with_updates(hashed_state.clone())?;
+        let mut state_root_result = None;
+        let persistence_in_progress = self.persistence_state.in_progress();
+        if !persistence_in_progress {
+            // NOTE: there is a possibility of a race condition here if some data was persisted. ???
```
Comment (rkrasiuk) [inline crates/engine/tree/src/tree/mod.rs:1881]:
```suggestion
```

File: crates/engine/tree/src/tree/mod.rs [hunk 2]
```diff
@@ -2167,8 +2170,29 @@ where
         let hashed_state = HashedPostState::from_bundle_state(&output.state.state);
 
         let root_time = Instant::now();
-        let (state_root, trie_output) =
-            state_provider.state_root_with_updates(hashed_state.clone())?;
+        let mut state_root_result = None;
+        let persistence_in_progress = self.persistence_state.in_progress();
+        if !persistence_in_progress {
```
Comment (mattsse) [inline crates/engine/tree/src/tree/mod.rs:2175]: let's inline this  and get rid of the tmp var

File: crates/engine/tree/src/tree/mod.rs [hunk 3]
```diff
@@ -2167,8 +2170,29 @@ where
...
+                Err(error) => return Err(error.into()),
+            };
+        }
+
+        let (state_root, trie_output) = match state_root_result {
+            Some(result) => result,
+            None => {
+                debug!(target: "engine", persistence_in_progress, "Failed to compute state root in parallel");
+                state_provider.state_root_with_updates(hashed_state.clone())?
+            }
```
Comment (mattsse) [inline crates/engine/tree/src/tree/mod.rs:2193]: pedantic style nit: if let some

File: crates/engine/tree/src/tree/mod.rs (hunk 2 above)
Comment (mattsse) [inline crates/engine/tree/src/tree/mod.rs:2175]: can we add some docs here why this is sound

**Review by mattsse — CHANGES_REQUESTED** [review]
nice,
pedantic style and doc nits

File: crates/engine/tree/src/tree/mod.rs (hunk 2 above)
Comment (rkrasiuk) [reply crates/engine/tree/src/tree/mod.rs:2175]: temp var cuz it's used later for logging

**Review by mattsse — APPROVED** [review]

**Pull request merged**